### PR TITLE
fix + optimization

### PR DESCRIPTION
**Pull request by Supercip971**

### CODE REVIEW

File: userspace/libraries/libc/string.cpp [hunk 1]
```diff
@@ -62,10 +62,37 @@ void *memmove(void *dest, const void *src, size_t n)
     return dest;
 }
 
+#ifdef __x86_64__
```
Comment (sleepy-monax) [inline userspace/libraries/libc/string.cpp:65]: GCC is really smart and generates nearly the same code for the two functions by using 128bit MMX registers internally ;)

File: userspace/libraries/libc/string.cpp [hunk 2]
```diff
@@ -62,10 +62,37 @@ void *memmove(void *dest, const void *src, size_t n)
...
+    return s1;
+}
+
+#else
+
 void *memcpy(void *s1, const void *s2, size_t n)
 {
     unsigned int *ldest = (unsigned int *)s1;
-    unsigned int *lsrc = (unsigned int *)s2;
+    const unsigned int *lsrc = (const unsigned int *)s2;
```
Comment (sleepy-monax) [inline userspace/libraries/libc/string.cpp:95]: yay more const :+1:

**Comment by sleepy-monax** [comment]
When doing "optimization" PRs you need to provide real-world numbers to back it up

**Comment by Supercip971** [comment]
benchmark with differents memcpy code:
```
Run on (12 X 5153.03 MHz CPU s)
CPU Caches:
  L1 Data 32 KiB (x6)
  L1 Instruction 32 KiB (x6)
  L2 Unified 512 KiB (x6)
  L3 Unified 16384 KiB (x2)
Load Average: 0.92, 0.75, 0.60
***WARNING*** CPU scaling is enabled, the benchmark real time measurements may be noisy and will incur extra overhead.
-----------------------------------------------------------------------------------
Benchmark                         Time             CPU   Iterations UserCounters...
-----------------------------------------------------------------------------------
memcpy_8_benchmark/8           2.70 ns         2.69 ns    260116061 bytes_per_second=2.7688G/s
memcpy_8_benchmark/16          4.81 ns         4.81 ns    145692926 bytes_per_second=3.09803G/s
memcpy_8_benchmark/32          10.3 ns         10.3 ns     68435643 bytes_per_second=2.89169G/s
memcpy_8_benchmark/64          21.5 ns         21.5 ns     33483402 bytes_per_second=2.77434G/s
memcpy_8_benchmark/128         38.3 ns         38.3 ns     18500586 bytes_per_second=3.11639G/s
memcpy_8_benchmark/256         70.3 ns         70.2 ns      9683457 bytes_per_second=3.39487G/s
memcpy_8_benchmark/512          132 ns          132 ns      4876485 bytes_per_second=3.61367G/s
memcpy_8_benchmark/1024         309 ns          309 ns      2238157 bytes_per_second=3.08467G/s
memcpy_8_benchmark/2048         652 ns          651 ns      1066162 bytes_per_second=2.92883G/s
memcpy_8_benchmark/4096        1349 ns         1347 ns       525013 bytes_per_second=2.83228G/s
memcpy_8_benchmark/8192        2710 ns         2707 ns       258974 bytes_per_second=2.81873G/s

memcpy_32_benchmark/8          1.61 ns         1.61 ns    441485129 bytes_per_second=4.63141G/s
memcpy_32_benchmark/16         2.61 ns         2.61 ns    268788976 bytes_per_second=5.71789G/s
memcpy_32_benchmark/32         5.14 ns         5.13 ns    136541499 bytes_per_second=5.80461G/s
memcpy_32_benchmark/64         14.0 ns         14.0 ns     50340243 bytes_per_second=4.25629G/s
memcpy_32_benchmark/128        24.7 ns         24.7 ns     28486342 bytes_per_second=4.83564G/s
memcpy_32_benchmark/256        41.8 ns         41.8 ns     16697822 bytes_per_second=5.70399G/s
memcpy_32_benchmark/512        69.3 ns         69.3 ns     10241142 bytes_per_second=6.88228G/s
memcpy_32_benchmark/1024        131 ns          131 ns      5279556 bytes_per_second=7.27227G/s
memcpy_32_benchmark/2048        258 ns          257 ns      2716812 bytes_per_second=7.40989G/s
memcpy_32_benchmark/4096        510 ns          510 ns      1374831 bytes_per_second=7.48454G/s
memcpy_32_benchmark/8192       1012 ns         1011 ns       690983 bytes_per_second=7.54428G/s

memcpy_64_benchmark/8          1.23 ns         1.23 ns    568051701 bytes_per_second=6.05299G/s
memcpy_64_benchmark/16         1.58 ns         1.58 ns    441773922 bytes_per_second=9.41246G/s
memcpy_64_benchmark/32         2.59 ns         2.58 ns    269051191 bytes_per_second=11.5332G/s
memcpy_64_benchmark/64         4.71 ns         4.71 ns    148522592 bytes_per_second=12.6562G/s
memcpy_64_benchmark/128        8.64 ns         8.62 ns     80780700 bytes_per_second=13.8263G/s
memcpy_64_benchmark/256        17.0 ns         16.9 ns     41387192 bytes_per_second=14.0966G/s
memcpy_64_benchmark/512        32.7 ns         32.6 ns     21464394 bytes_per_second=14.614G/s
memcpy_64_benchmark/1024       69.5 ns         69.3 ns     10229600 bytes_per_second=13.7651G/s
memcpy_64_benchmark/2048        133 ns          132 ns      5370837 bytes_per_second=14.4351G/s
memcpy_64_benchmark/4096        260 ns          259 ns      2733700 bytes_per_second=14.7182G/s
memcpy_64_benchmark/8192        509 ns          509 ns      1375733 bytes_per_second=14.9903G/s
```
on g++ -O2
(with g++ -O3 there is not a lot of differences)

benchmark memcpy source code:
```c++
#include <benchmark/benchmark.h>
void *memcpy64(void *s1, const void *s2, size_t n)
{
    int64_t *ldest = (int64_t *)s1;
    const int64_t*lsrc = (const int64_t *)s2;

    while (n >= sizeof(int64_t))
    {
        *ldest++ = *lsrc++;
        n -= sizeof(int64_t);
    }

    char *cdest = (char *)ldest;
    const char *csrc = (const char *)lsrc;

    while (n > 0)
    {
        *cdest++ = *csrc++;
        n -= 1;
    }

    return s1;
}
void *memcpy32(void *s1, const void *s2, size_t n)
{
    int32_t *ldest = (int32_t *)s1;
    const int32_t*lsrc = (const int32_t *)s2;

    while (n >= sizeof(int32_t))
    {
        *ldest++ = *lsrc++;
        n -= sizeof(int32_t);
    }

    char *cdest = (char *)ldest;
    const char *csrc = (const char *)lsrc;

    while (n > 0)
    {
        *cdest++ = *csrc++;
        n -= 1;
    }

    return s1;
}

void *memcpy8(void *s1, const void *s2, size_t n)
{

    char *cdest = (char *)s1;
    const char *csrc = (const char *)s2;

    while (n > 0)
    {
        *cdest++ = *csrc++;
        n -= 1;
    }

    return s1;
}

static void memcpy_8_benchmark(benchmark::State& state) {
  // Code inside this loop is measured repeatedly
  
  uint8_t* temp_data1 = new uint8_t[state.range(0)+512];
  uint8_t* temp_data2 = new uint8_t[state.range(0)+512];
  
  for (auto _ : state) {
    memcpy8(temp_data1, temp_data2, state.range(0));

    benchmark::DoNotOptimize(temp_data1);
    benchmark::DoNotOptimize(temp_data2);
  }

  delete[] temp_data1;
  delete[] temp_data2;
  state.SetBytesProcessed(int64_t(state.iterations()) *
                          int64_t(state.range(0)));

}
BENCHMARK(memcpy_8_benchmark)->RangeMultiplier(2)->Range(8, 8<<10);

static void memcpy_32_benchmark(benchmark::State& state) {
  // Code inside this loop is measured repeatedly
  
  uint8_t* temp_data1 = new uint8_t[state.range(0)+512];
  uint8_t* temp_data2 = new uint8_t[state.range(0)+512];
  
  for (auto _ : state) {
    memcpy32(temp_data1, temp_data2, state.range(0));

    benchmark::DoNotOptimize(temp_data1);
    benchmark::DoNotOptimize(temp_data2);
  }
  delete[] temp_data1;
  delete[] temp_data2;

  state.SetBytesProcessed(int64_t(state.iterations()) *
                          int64_t(state.range(0)));
  
}
BENCHMARK(memcpy_32_benchmark)->RangeMultiplier(2)->Range(8, 8<<10);

static void memcpy_64_benchmark(benchmark::State& state) {
  // Code inside this loop is measured repeatedly
  
  uint8_t* temp_data1 = new uint8_t[state.range(0)+512];
  uint8_t* temp_data2 = new uint8_t[state.range(0)+512];
  
  for (auto _ : state) {
    memcpy64(temp_data1, temp_data2, state.range(0));

    benchmark::DoNotOptimize(temp_data1);
    benchmark::DoNotOptimize(temp_data2);
  }
  delete[] temp_data1;
  delete[] temp_data2;

  state.SetBytesProcessed(int64_t(state.iterations()) *
                          int64_t(state.range(0)));
  
}
BENCHMARK(memcpy_64_benchmark)->RangeMultiplier(2)->Range(8, 8<<10);

BENCHMARK_MAIN();
```

File: userspace/libraries/libc/string.cpp [hunk 3]
```diff
@@ -62,42 +62,25 @@ void *memmove(void *dest, const void *src, size_t n)
...
-        *cdest++ = *csrc++;
-        n -= 1;
-    }
-
-    return s1;
-}
 
+    // use the best size for memcpy block depending on the architecture
+#ifdef __x86_64__
+    using MemcpyBlock = int64_t;
```
Comment (feliwir) [inline userspace/libraries/libc/string.cpp:70]: this is identical to `intptr_t` defined by stdint.h

**Comment by sleepy-monax** [comment]
ok looks good :+1:

**Pull request merged**